language/return_spec.rb

```diff
@@ -249,4 +249,202 @@ def f
+            while true
+              return
+            end
+
+            puts "after while"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+      end
+
+      describe "within a begin" do
+        it "is allowed in begin block" do
+          ruby_exe(<<-END_OF_CODE).should == "before begin\n"
+            puts "before begin"
+            begin
+              return
+            end
+            puts "after begin"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+
+        it "is allowed in ensure block" do
+          ruby_exe(<<-END_OF_CODE).should == "before begin\n"
+            puts "before begin"
+            begin
+            ensure
+              return
+            end
+            puts "after begin"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+
+        it "is allowed in rescue block" do
+          ruby_exe(<<-END_OF_CODE).should == "before begin\n"
+            puts "before begin"
+            begin
+              raise
+            rescue RuntimeError
+              return
+            end
+            puts "after begin"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+
+        it "fires ensure block before returning" do
+          ruby_exe(<<-END_OF_CODE).should == "within ensure\n"
+            begin
+              return
+            ensure
+              puts "within ensure"
+            end
+
+            puts "after begin"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+
+        it "swallows exception if returns in ensure block" do
+          ruby_exe(<<-END_OF_CODE).should == ""
+            begin
+              raise
+            ensure
+              return
+            end
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+      end
+
+      describe "within a block" do
+        it "is allowed" do
+          ruby_exe(<<-END_OF_CODE).should == "before call\n"
+            puts "before call"
+            proc { return }.call
+
+            puts "after call"
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+      end
+
+      describe "within a class" do
+        it "is allowed" do
+          ruby_exe(<<-END_OF_CODE).should == "before return\n"
+            class A
+              puts "before return"
+              return
+
+              puts "after return"
+            end
+          END_OF_CODE
+
+          $?.exitstatus.should == 0
+        end
+      end
+
+      describe "file loading" do
+        it "stops file loading and execution" do
+          require "tempfile"
+
+          temp = Tempfile.new(["loaded_file", ".rb"])
+          temp.write <<-END_OF_CODE
+            puts "before return"
+            return
+            puts "after return"
+          END_OF_CODE
+          temp.flush
+
+          ruby_exe(<<-END_OF_CODE).should == "before return\n"
+            load "#{temp.path}"
+          END_OF_CODE
```

Comment:
The file needs to be removed after the end of the spec.
It's also better to use the `tmp("loaded_file.rb")` and `rm_r` helpers as they avoid creating any file outside the rubyspec_temp directory.